clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -445,54 +534,126 @@ private void updateGroupSubscription(Set<String> topics) {
+            final TopicPartition tp = entry.getKey();
+            if (!assignedPartitions.contains(tp)) {
+                assignments.get(entry.getValue()).partitions().add(tp);
+            }
+        }
+
+        // if revocations are triggered, tell everyone to re-join immediately.
+        if (revocationsNeeded) {
+            for (final Assignment assignment : assignments.values()) {
+                assignment.setError(ConsumerProtocol.Errors.NEED_REJOIN);
+            }
+        }
+    }
+
     @Override
     protected void onJoinPrepare(int generation, String memberId) {
         // commit offsets prior to rebalance if auto-commit enabled
         maybeAutoCommitOffsetsSync(time.timer(rebalanceTimeoutMs));
 
         // execute the user's callback before rebalance
         ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
-        Set<TopicPartition> revoked = subscriptions.assignedPartitions();
-        log.info("Revoking previously assigned partitions {}", revoked);
-        try {
-            listener.onPartitionsRevoked(revoked);
-        } catch (WakeupException | InterruptException e) {
-            throw e;
-        } catch (Exception e) {
-            log.error("User provided listener {} failed on partition revocation", listener.getClass().getName(), e);
+
+        switch (protocol) {
+            case EAGER:
+                Set<TopicPartition> revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                log.info("Revoking previously assigned partitions {}", revokedPartitions);
+                try {
+                    listener.onPartitionsRevoked(revokedPartitions);
+                } catch (WakeupException | InterruptException e) {
+                    throw e;
+                } catch (Exception e) {
+                    log.error("User provided listener {} failed on partition revocation", listener.getClass().getName(), e);
+                }
+
+                // also clear the assigned partitions since all have been revoked
+                subscriptions.assignFromSubscribed(Collections.emptySet());
+
+                break;
+
+            case COOPERATIVE:
+                break;
         }
 
         isLeader = false;
         subscriptions.resetGroupSubscription();
     }
 
+    @Override
+    public void resetGeneration(String errorMessage) {
+        maybeLosePartitions(errorMessage);
+        super.resetGeneration(errorMessage);
+    }
+
     @Override
     public boolean rejoinNeededOrPending() {
         if (!subscriptions.partitionsAutoAssigned())
             return false;
 
         // we need to rejoin if we performed the assignment and metadata has changed
-        if (assignmentSnapshot != null && !assignmentSnapshot.matches(metadataSnapshot))
+        if (assignmentSnapshot != null && !assignmentSnapshot.matches(metadataSnapshot)) {
+            maybeLosePartitions("topic metadata has changed and therefore some topics may not exist any moreq");
```

Comment:
typo: moreq -> more

Reply:
Also: should only call onPartitionsLost on owned partitions that no longer exist

Reply:
Good catch.